### ISSUE TITLE
RFC: Requirement analysis v1

### ISSUE THREAD

**Issue by lakhansamani** [promise-learning/code-playground#32]
This RFC is created to discuss the requirements for v1.
Here we are trying to a better version of JS Fiddle. Before doing that we would like to analyze the existing products and see if this is a feasible product to make and why would someone choose it!

**Comment by lakhansamani** [comment]
## Existing Products


| Products⬇️ x Features ➡️ | Code Editor  | Transpiler  | Module Support | Console | Live Preview | Live code sharing | Git Integration | Themeing | File Explorer | Hosting | multiple sandboxes
| ---  | --- | ---  | --- | --- | --- |--- | --- | --- | --- | --- | --- |
| CodeSandbox | ✅ | ✅  | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ |
| Codepad  | |  | | | | |  |  |  |  |  |
| ReplIt  | ✅ | ✅  | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ |
  

## What are we doing differently?

## Why we need this product?

## Why would people choose us?

## Use cases
* Testing JS code snippet
* Creating an interview platform [ Just an idea here, can we create an npm module which interviewers can use to create white label solutions]. This can help interviewers create their own interview sessions and platforms in minutes, without paying huge amounts to a platform like a hacker rank.

## Learnings

* UX improvement
* Building product with live socket and live editing
* How npm modules can be installed in a browser which is not in UMD formats
* WASM
* Test cases and strategies
* AST exploration

**Comment by lakhansamani** [comment]
@jyash97 can you put your case study here!

**Comment by lakhansamani** [comment]
Thinking aloud here!

can we have something like a [spectrum](https://github.com/withspectrum/spectrum)
but it should be focused on creating tests/exercises which can be shared further.
also, you can deploy your own version of it.

**Comment by nikhil-varma** [comment]
@lakhansamani Nice one! This can be a reference for us to make sure our pipeline is in the right direction. Also this will be an Epic and not a story :P

**Comment by lakhansamani** [comment]
Right, let's find out a niche requirement and build something around that!

**Comment by nikhil-varma** [comment]
### Existing competitors
- CodeSandbox
- CoderPad
- Repl.it
- CodeShare
- Stackblitz
- Glitch
- JSFiddle

## Feature set
- Multi-lingual coding system (**_MVP_**)
  - JavaScript
  - HTML
  - CSS/SCSS/LESS
  - FE Frameworks
  - Node.js
  - Next.js
  - Gatsby
  - OpenFaaS supporting languages
    - Java
    - C/C++
    - Python
- IDE (**_MVP_**)
- Embed (**_MVP_**)
- CI systems (_Integrate-able_)
- Collaborations 
   - Teams
      - Code snippets and templates 
   - Interviews 
- Integrations
   - Github integrations
      - Sandbox to Github and vice versa
- Deployments and CD systems
   - Netlify
- Containerisation
    - Docker
    - Kubernetes
- Local & personalised instances
- Seamless interview experience
   - Live collaboration
   - Pre-made tests
   - Timer system (_integrate-able_)
- Assertion tools (_For multiple languages_)  (**_MVP_**)
- Developer tools (**_MVP_**)
- Reviewer system
- OpenAPI support for any ecosystem  (**_MVP_**)
